src/hardware/voodoo.cpp

```diff
@@ -34,155 +37,171 @@
 #include "voodoo_interface.h"
 
 
-class VOODOO;
-static VOODOO* voodoo_dev;
+class Voodoo;
+static std::unique_ptr<Voodoo> voodoo_dev;
+
+constexpr uint32_t voodoo_lfb_mask = 0xffff0000;
+static uint32_t voodoo_current_lfb = (VOODOO_INITIAL_LFB & voodoo_lfb_mask);
 
-static uint32_t voodoo_current_lfb=(VOODOO_INITIAL_LFB&0xffff0000);
+static_assert((VOODOO_INITIAL_LFB & 0xffff0000) == VOODOO_INITIAL_LFB,
+              "VOODOO_INITIAL_LFB must have its lower 16 bits set to zero");
 
+enum class EmulationType {
+    None,
+    Software,
+    OpenGL
+};
 
-class VOODOO:public Module_base{
+enum class CardType : int {
+    Type1 = 1
+};
+
+constexpr std::string_view NoneStr = "false";
+constexpr std::string_view SoftwareStr = "software";
+constexpr std::string_view OpenGLStr = "opengl";
+constexpr std::string_view AutoStr = "auto";
+
+class Voodoo {
 private:
-	Bits emulation_type;
-public:
-	VOODOO(Section* configuration):Module_base(configuration){
-		emulation_type=-1;
+    Section* m_configuration;
+    EmulationType emulation_type;
 
-		Section_prop * section=static_cast<Section_prop *>(configuration);
-		std::string voodoo_type_str(section->Get_string("voodoo_card"));
-		if (voodoo_type_str=="false") {
-			emulation_type=0;
-		} else if (voodoo_type_str=="software" || (voodoo_type_str=="auto" && !OpenGL_using())) {
-			emulation_type=1;
-#if C_OPENGL
-		} else if (voodoo_type_str=="opengl" || voodoo_type_str=="auto") {
-			emulation_type=2;
-#else
-		} else if (voodoo_type_str=="auto") {
-			emulation_type=1;
-#endif
+public:
+    Voodoo(const Voodoo&) = delete;
+    Voodoo& operator=(const Voodoo&) = delete;
+	
+    Voodoo(Section* configuration) : m_configuration(configuration), emulation_type(EmulationType::None) {
+        UpdateConfiguration(configuration);
+    }
+	
+	EmulationType GetEmulationTypeFromString(std::string_view voodoo_type_str) {
+		if (voodoo_type_str == NoneStr) {
+			return EmulationType::None;
+		} else if (voodoo_type_str == SoftwareStr) {
+			return EmulationType::Software;
+		} else if (voodoo_type_str == OpenGLStr) {
+			return C_OPENGL ? EmulationType::OpenGL : EmulationType::None;
+		} else if (voodoo_type_str == AutoStr) {
+			return C_OPENGL ? EmulationType::OpenGL : EmulationType::Software;
 		} else {
-			emulation_type=0;
+			return !OpenGL_using() ? EmulationType::Software : EmulationType::None;
```

Comment:
This is not something added by you, but can we rename `OpenGL_using` to something normal? 😄 lower_snake_case, for a start, and should be probably `is_using_opengl` or something.

Reply:
I agree!